source/util/EnvUtils.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "EnvUtils.h"
+#include "../logging/LoggerFactory.h"
+
+#include <cerrno>
+#include <cstdlib>
+#include <sstream>
+#include <vector>
+
+#include <linux/limits.h>
+#include <string.h>
+#include <unistd.h>
+
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+char *OSInterfacePosix::getenv(const char *name)
+{
+    return ::getenv(name);
+}
+
+int OSInterfacePosix::setenv(const char *name, const char *value, int overwrite)
+{
+    return ::setenv(name, value, overwrite);
+}
+
+char *OSInterfacePosix::getcwd(char *buf, size_t size)
+{
+    return ::getcwd(buf, size);
+}
+
+// Sequences of path prefixes used to search for executable filenames.
+constexpr char PATH_ENVIRONMENT[] = "PATH";
+
+// Separator between path prefixes in environment variable.
+constexpr char PATH_ENVIRONMENT_SEPARATOR = ':'; // Unix-only.
+
+// Separator between directories in path.
+constexpr char PATH_DIRECTORY_SEPARATOR = '/'; // Unix-only.
+
+// Jobs directory name.
+constexpr char JOBS_DIRECTORY_NAME[] = "jobs";
+
+constexpr char EnvUtils::TAG[];
+
+int EnvUtils::AppendCwdToPath()
+{
+    // Use a vector of char as buffer for path to current working directory.
+    //
+    // Maximum number of characters in absolute path is filesystem dependent.
+    //
+    // The getcwd function takes a buffer and length and returns the current
+    // working directory or NULL if the length of the buffer is insufficient
+    // to store the current working directory.
+    //
+    // As a result, we initialize the buffer with the maximum length of a
+    // relative pathname (_PC_PATH_MAX) and then repeatedly resize the buffer
+    // in the event that getcwd is unable to fit the current working directory
+    // in the buffer passed as input.
+    std::vector<char> cwd;
+    long path_max = pathconf(".", _PC_PATH_MAX);
+    if (path_max < 1)
+    {
+        cwd.resize(PATH_MAX + 1, '\0'); // Fallback to hardcoded PATH_MAX.
+    }
+    else
+    {
+        cwd.resize(path_max + 1, '\0');
+    }
+
+    char *pcwd = nullptr;
+    while (!pcwd)
```

Comment:
We should avoid having infinite loop. Considering the PATH_MAX value for most Linux OS, lets end this look with maximum 3 retries.

Reply:
Done.  We will resize at most 3 times from initial size. Unit test added to exercise this failure mode.